pkg/webui/lib/errors/utils_test.js

```diff
@@ -0,0 +1,225 @@
+// limitations under the License.
+
+import { withScope, captureException } from '@sentry/browser'
+
+import { getSentryErrorTitle, createFrontendError, ingestError } from './utils'
+import errorMessages from './error-messages'
+
+jest.mock('@sentry/browser')
+jest.mock('@ttn-lw/lib/log')
+
+const setTags = jest.fn()
+const setExtras = jest.fn()
+const setFingerprint = jest.fn()
+
+captureException.mockImplementation(jest.fn)
+withScope.mockImplementation(callback => callback({ setTags, setExtras, setFingerprint }))
+
+const backendError = {
+  code: 2,
+  message:
+    'error:pkg/assets:http (HTTP error: `` is not a valid ID. Must be at least 2 and at most 36 characters long and may consist of only letters, numbers and dashes. It may not start or end with a dash)',
+  details: [
+    {
+      '@type': 'type.googleapis.com/ttn.lorawan.v3.ErrorDetails',
+      namespace: 'pkg/assets',
+      name: 'http',
+      message_format: 'HTTP error: {message}',
+      attributes: {
+        message:
+          '`` is not a valid ID. Must be at least 2 and at most 36 characters long and may consist of only letters, numbers and dashes. It may not start or end with a dash',
+      },
+    },
+  ],
+}
+
+const conflictBackendError = {
+  code: 6,
+  message: 'error:pkg/identityserver/store:id_taken (ID already taken)',
+  details: [
+    {
+      '@type': 'type.googleapis.com/ttn.lorawan.v3.ErrorDetails',
+      namespace: 'pkg/identityserver/store',
+      name: 'id_taken',
+      message_format: 'ID already taken',
+      correlation_id: '06d95b39becd435cbb67ab87a3c93312',
+      code: 6,
+    },
+  ],
+  request_details: { url: '/users/kschiffer/applications', method: 'post', stack_component: 'is' },
+}
+
+const frontendError = createFrontendError(
+  errorMessages.unknownErrorTitle,
+  undefined,
+  undefined,
+  500,
+)
+const plainFrontendError = createFrontendError(errorMessages.unknownErrorTitle)
+const codeError = { code: 'ECONNABORTED' }
+const statusCodeError = { statusCode: 404 }
+const emptyError = {}
+const undefinedError = undefined
+const errorInstance = new Error('There was an unknown error')
+
+describe('Get Sentry error title', () => {
+  it('retrieves the right error title', () => {
+    expect(getSentryErrorTitle(backendError)).toBe(backendError.message)
+    expect(getSentryErrorTitle(frontendError)).toBe(frontendError.errorTitle.defaultMessage)
+    expect(getSentryErrorTitle(codeError)).toBe(codeError.code)
+    expect(getSentryErrorTitle(statusCodeError)).toBe(`status code: ${statusCodeError.statusCode}`)
+    expect(getSentryErrorTitle(emptyError)).toBe('untitled or empty error')
+    expect(getSentryErrorTitle(undefinedError)).toBe('invalid error type')
+  })
+})
+
+describe('Ingest error', () => {
+  beforeEach(() => {
+    jest.clearAllMocks()
+  })
+
+  describe('when passing backend errors', () => {
+    it('correctly forwards sentry-worthy errors', () => {
+      ingestError(backendError, undefined, undefined, () => {
```

Comment:
What is the purpose of making all assertions in the `done` callback if `ingestError` and `withScope` are synchronous? Not sure about the actual implementation of `withScope` but the mocked version is definitely synchronous. If you put all assertion right after the call to `ingestError` everything should work just fine.